framework/include/userobject/BulkMaterial.h

```diff
@@ -0,0 +1,242 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "ElementUserObject.h"
+
+#include "libmesh/int_range.h"
+
+#include <tuple>
+#include <string>
+#include <vector>
+#include <utility>
+
+// type wrappers
+template <typename T>
+struct GatherMatProp
+{
+  typedef T type;
+  typedef const MaterialProperty<T> ref;
+
+  template <typename M>
+  static ref * getPointer(M & mpi, const MaterialPropertyName & name)
+  {
+    return &mpi.template getMaterialProperty<T>(name);
+  }
+};
+
+struct GatherVariable
+{
+  typedef Real type;
+  typedef const VariableValue ref;
+
+  template <typename C>
+  static ref * getPointer(const C & coupleable, const VariableName & name)
+  {
+    return &coupleable.coupledValue(name);
+  }
+};
+
+// tuple wrappers
+struct TupleStd
+{
+  // the tuple type
+  template <typename... Args>
+  using type = std::tuple<Args...>;
+
+  // the element access function
+  template <int N, typename T>
+  static auto & get(T && t)
+  {
+    return std::get<N>(t);
+  }
+
+  // tuple size (number of elements)
+  template <typename T>
+  using size = std::tuple_size<T>;
+
+  // element type access
+  template <int I, typename T>
+  using element = typename std::tuple_element<I, T>::type;
+};
+
+// this wrapper is yet untested
+#ifdef CUDA_SUPPORTED
+struct TupleCuda
+{
+  // the tuple type
+  template <typename... Args>
+  using type = cuda::std::tuple<Args...>;
+
+  // the element access function
+  template <int N, typename T>
+  static auto & get(T && t)
+  {
+    return cuda::std::get<N>(t);
+  }
+
+  // tuple size (number of elements)
+  template <typename T>
+  using size = cuda::std::tuple_size<T>;
+
+  // element type access
+  template <int I, typename T>
+  using element = typename cuda::std::tuple_element<I, T>::type;
+};
+#endif
+
+template <typename Tuple, typename Output, typename... Input>
+class BulkMaterial : public ElementUserObject
+{
+  /// shorthand for the instantiated template type
+  typedef BulkMaterial<Tuple, Output, Input...> BulkMaterialType;
+
+  /// Helper template for obtainig all variable and material property references
+  template <int I, typename T, typename... Names>
+  void construct(T && name, Names &&... names)
```

Comment:
In general I don't like having multi-line method definitions in class definitions as it makes it harder for a developer to peruse APIs